src/coreclr/vm/threads.h

```diff
@@ -1296,6 +1312,15 @@ class Thread
         return HasThreadStateNC(Thread::TSNC_EtwStackWalkInProgress);
     }
 
+    BOOL RequireThreadStoreLock()
```

Comment:
I have looked more into where this is used: The background GC threads that will have it set can be created and destroyes while the program is running. It is still not clear to me what makes it safe to skip the threadstore lock for them. What prevents some other thread from being added to `s_pThreadStore->m_ThreadList` at the same time the GC thread is being added to `s_pThreadStore->m_ThreadList`?

(It is possible that the existing logic had a latent bug that does not happen often enough to get noticed.)

Reply:
>It is still not clear to me what makes it safe to skip the threadstore lock for them. What prevents some other thread from being added to s_pThreadStore->m_ThreadList at the same time the GC thread is being added to s_pThreadStore->m_ThreadList?

Right. I am still catching up but even the 30 minutes I took last night confirmed there is a fair bit of complexity in this area. The implementation of the lock/unlock calls out to the `ThreadSuspend` class and I've never used this class at all. I'm going to need to take a few hours to debug/read in order to answer this question.

This is what I get for touching things 😄

Reply:
@jkotas I agree this name is bad because there are places where the BGC thread can get into trouble so I am going to rename it entirely. Note that only the BGC has this logic so lets focus on how that scenario works. At present, there are 3 times when the new `RequireThreadStoreLock()` API is called.

1) `ThreadStore::AddThread()`
1) `Thread::SetBackground()`
1) `ThreadStore::TransferStartedThread()`

Scenarios (1) and (2) are fine because they are called in a gen2 GC when the BGC is first launched. This, based on my understanding means that no other threads are running so it is okay to do this. Cases when there could be other threads, I believe Server GC, are guarded because this create call is done within a `join` clause in the GC and with synchronization around the `prepare_bgc_thread()` method.

Scenario (3) is a bit more of an issue. This is called by the newly created BGC thread which means it can be running with other managed threads and as pointed out we could get into trouble. The beginning of this is in `HasStarted()`.

I am inclined to update the API name, conditionally take the lock for (1) and (2) based on BGC, and always take the lock in (3).

Reply:
Stand by. I just tried to let (3) take the lock and now I've got a deadlock. Investigating...

Reply:
> This, based on my current understanding, means that no other threads are running so it is okay to do this

It means that there are no threads running in cooperative mode. There can be other threads running in preemptive mode.

> conditionally take the lock for (1) and (2) based on BGC, and always take the lock in (3).

I guess we avoid taking the threadstore lock in (1) and (2) because of it is taken already, and taking it a second time would cause asserts or crashes. If it is the case, would it make sense to call the argument `fThreadStoreLockAlreadyTaken` and then assert in the implementation that the thread store lock is really taken when this argument is true.

Also, the problem with SetBackground thread can be solved by creating all threads as background by default, and only call this API in the minority places that create foreground threads. The foreground threads are the outlier that needs an extra work. Background threads are the common case.

Reply:
Doh. The (3) is actually for the same reason as (1) and (2). The creator thread is waiting for the BGC to get to a certain point before continuing. This effectively lets the BGC safely start up without the lock and then never needs it again. Basically what we have here is correct so no bug - yay. However, the entire naming situation here is confusing. I will take the `fThreadStoreLockAlreadyTaken` suggestion because that is the correct wording in this case.

After that I will look into the `SetBackground()` suggestion. That appears more complicated given the number of places that is called and not called so will require some more auditing.

Reply:
@jkotas I've applied the "already taken" concept and this makes things much clearer. Moving onto the `SetBackground()` suggestion.

Reply:
Deferring any additional cleanup to another PR.